btf/core.go

```diff
@@ -1045,13 +1050,295 @@ func coreAreMembersCompatible(localType Type, targetType Type) error {
+ *   - number and position of arguments in local type has to match target
+ *   - for each argument and the return value we recursively check match
+ */
+func coreTypesMatch(localType Type, targetType Type, behindPtr bool, visited map[pair]struct{}) error {
+	localType = UnderlyingType(localType)
+	targetType = UnderlyingType(targetType)
+
+	if !coreNamesMatch(localType.TypeName(), targetType.TypeName()) {
+		return fmt.Errorf("type name %q don't match %q: %w", localType.TypeName(), targetType.TypeName(), errIncompatibleTypes)
+	}
+
+	if _, ok := visited[pair{localType, targetType}]; ok {
+		return nil
+	}
+	if visited == nil {
+		visited = make(map[pair]struct{})
+	}
+	visited[pair{localType, targetType}] = struct{}{}
+
+	switch lv := (localType).(type) {
+	case *Void:
+		if _, ok := targetType.(*Void); !ok {
+			return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+		}
+
+	case *Fwd:
+		if behindPtr {
+			if tv, ok := targetType.(*Fwd); ok {
+				if lv.Kind != tv.Kind {
+					return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+				}
+
+				return nil
+			}
+
+			if _, ok := targetType.(*Struct); ok && lv.Kind == FwdStruct {
+				return nil
+			}
+
+			if _, ok := targetType.(*Union); ok && lv.Kind == FwdUnion {
+				return nil
+			}
+
+			return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+		}
+
+		if tv, ok := targetType.(*Fwd); ok && tv.Kind == lv.Kind {
+			return nil
+		}
+
+		return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+
+	case *Enum:
+		tv, ok := targetType.(*Enum)
+		if !ok {
+			return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+		}
+
+		if err := coreEnumsMatch(lv, tv); err != nil {
+			return err
+		}
+
+	case *Struct, *Union:
+		if behindPtr {
+			if reflect.TypeOf(localType) == reflect.TypeOf(targetType) {
+				return nil
+			}
+
+			tv, ok := targetType.(*Fwd)
+			if !ok {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if _, ok := lv.(*Struct); ok && tv.Kind == FwdStruct {
+				return nil
+			}
+
+			if _, ok := lv.(*Union); ok && tv.Kind == FwdUnion {
+				return nil
+			}
+
+			return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+		}
+
+		if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
+			return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+		}
+
+		lComp, ok := lv.(composite)
+		if !ok {
+			return fmt.Errorf("expected composite type, got %T", lv)
+		}
+
+		tComp, ok := targetType.(composite)
+		if !ok {
+			return fmt.Errorf("expected composite type, got %T", targetType)
+		}
+
+		if err := coreCompositesMatch(lComp, tComp, behindPtr, visited); err != nil {
```

Comment:
Hmm, so again, due to the implementation in LibBPF. We now recuse instead of having the stacks or the goto loop in LibBPF. In LibBPF, the `behindPtr` defaults to false, is set to `true` once we dive into a pointer, but never is changed back to `false`. 

But I do agree with you that this doesn't seem to make sense. I found a selftest in libbpf for this logic, so I will port it and see if I can change this bit of the logic without breaking the LibBPF test.

Reply:
I added `TestCORETypesMatchSelftest` which is based on the selftests, it is used by libbpf to verify a number cases, having both positive and negative matches. It passed on the original code, I then changed `behindPtr` to `false` wherever possible so we only ever pass true to 1 descendant and the test keeps working.  We this change doesn't seem to violate any assumptions.